### ISSUE TITLE
TF: XLA generation not working properly in some models

### ISSUE THREAD

**Issue by gante** [huggingface/transformers#17935]
This issue is used to track TensorFlow XLA generation issues, arising from #17857. There are three categories of issues, sorted in descending order by severity:

### Key model issues

These are heavily-used models, whose quality should be prioritized.

- [x] T5 -- The quality of the results decreases with `max_length`. See [here](https://github.com/huggingface/transformers/pull/17857/files#r906702367).
- [x] GPT-J -- fails simple generate tests with numerical issues

### Models failing basic tests

These models are failing `test_xla_generate_fast` -- a short greedy generation.

- [ ] LED
- [ ] Speech2Text
- [ ] XLNet
- [ ] XGLM

### Models failing complex tests

These are models failing `test_xla_generate_slow` -- a long beam search generation.

- [x] Bart
- [x] Blenderbot
- [x] Marian
- [x] mbart
- [x] OPT
- [x] Pegasus

**Comment by anmolsjoshi** [comment]
@gante do you require any help with this issue? Happy to contribute

**Comment by gante** [comment]
Hi @anmolsjoshi 👋 

If you are comfortable with debugging XLA, absolutely :) My recommendation would be to pick a model from "Models failing complex tests" (the others might require significant architecture changes), and to start debugging. The number 1 suspect is always the position embeddings, which may not be handling the case where `past` is padded. Let me know if you are up to it, and which model would you like to take!

**Comment by dsuess** [comment]
Hi @gante, I did have a bit of a poke around. I think the complex tests all fail for the same reason: those models have a setting `max_position_embeddings`, which is set to 20 by default during testing. Here’s a simple fix for those: https://github.com/dsuess/transformers/commit/4a3e27164ae941fcd649b8565d7d92a4552d689f. I’ll give the other ones a shot now

**Comment by JuheonChu** [comment]
Hello @gante, may I ask if there is anything that I can contribute?

**Comment by gante** [comment]
Hi JuheonChu 👋 Actually yes! I have a few unchecked models at the top, but I wouldn't recommend spending time there unless you plan to use those architectures -- they are infrequently used.

However, two popular models are failing their XLA tests with beam search:
- Marian
- OPT

You can see the failing test if you install from `main` (`pip install --upgrade git+https://github.com/huggingface/transformers.git`) and run the slow test. e.g. for OPT `NVIDIA_TF32_OVERRIDE=0 RUN_SLOW=1 py.test -vv tests/models/opt/test_modeling_tf_opt.py::TFOPTModelTest::test_xla_generate_slow`

I don't know what's the cause for the failure, so you'll have to hop into debug mode and see what is breaking :)

**Comment by JuheonChu** [comment]
Can @katiele47 and I try working on them?

**Comment by gante** [comment]
@JuheonChu of course!

**Comment by JuheonChu** [comment]
> @JuheonChu of course!
@gante  May I ask you if we are figuring out the root of following errors when running commands for the OPT model?
![Error](https://user-images.githubusercontent.com/35699839/219705302-a02bda62-755f-4c34-8359-67bfc3a1dd93.png)

**Comment by gante** [comment]
@JuheonChu yes. My suggestion would be to attempt to find where the numerical differences start from (between the XLA and the non-XLA version), using a debugger. Please note that you can't print variables with `jit_compile=True`, so you should set it to `False`. From there, the root cause is typically apparent.

Be warned, these sort of tasks sometimes are very time-consuming to complete :)

**Comment by JuheonChu** [comment]
> @JuheonChu yes. My suggestion would be to attempt to find where the numerical differences start from (between the XLA and the non-XLA version), using a debugger. Please note that you can't print variables with `jit_compile=True`, so you should set it to `False`. From there, the root cause is typically apparent.
> 
> Be warned, these sort of tasks sometimes are very time-consuming to complete :)

Thank you very much for your valuable guidance! We will try and keep you updated!

**Comment by katiele47** [comment]
Hi @gante, I've attempted to reproduce the failed XLA test on the OPT model using your suggested commands. The cause of error I had was somehow different from @JuheonChu's. Would you be able to verify if the following is the expected failing test output? If not, I assume it could be due to my local repo. Thanks!
<img width="1015" alt="Screen Shot 2023-02-21 at 11 20 44 PM" src="https://user-images.githubusercontent.com/54815905/220521074-6ab355c4-fe0b-42b5-88fa-cbd15de82b8a.png">
<img width="1125" alt="Screen Shot 2023-02-21 at 11 21 24 PM" src="https://user-images.githubusercontent.com/54815905/220521160-946a03db-cf80-4fbc-818c-3be337df3983.png">
<img width="1135" alt="Screen Shot 2023-02-21 at 11 21 43 PM" src="https://user-images.githubusercontent.com/54815905/220521210-981d6099-e093-4b8b-9913-2dafe5bef905.png">

**Comment by soma2000-lang** [comment]
@gante working on  XLNet

**Comment by qgallouedec** [comment]
TensorFlow and Flax were fully removed library-wide in #40760, shipped in v5.0.0, so the `modeling_tf_*.py` files and the `test_xla_generate_fast` / `test_xla_generate_slow` tests this issue tracks are gone from main.

Closing as outdated. Thanks to everyone who picked up a model here.
